lib/protocol/protocol.go

```diff
@@ -450,59 +450,59 @@ func (c *rawConnection) dispatcherLoop() (err error) {
 		case *Index:
 			l.Debugln("read Index message")
 			if state != stateReady {
-				return fmt.Errorf("protocol error: index message in state %d", state)
+				return indexError(fmt.Errorf("invalid state %d", state), "index", msg.Folder)
 			}
 			if err := checkIndexConsistency(msg.Files); err != nil {
-				return errors.Wrap(err, "protocol error: index")
+				return indexError(err, "index", msg.Folder)
 			}
 			if err := c.handleIndex(*msg); err != nil {
-				return fmt.Errorf("receiving index: %w", err)
+				return indexError(err, "index", msg.Folder)
```

Comment:
The error returns from `handle...` should not be wrapped as `protocol error: ...` I think